clippy_lints/src/loops.rs

```diff
@@ -566,6 +566,17 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Loops {
             ) = (pat, &match_expr.kind)
             {
                 let iter_expr = &method_args[0];
+
+                // Don't lint when the iterator is recreated on every iteration
+                if_chain! {
+                    if let ExprKind::MethodCall(..) | ExprKind::Call(..) = iter_expr.kind;
+                    if let Some(iter_def_id) = get_trait_def_id(cx, &paths::ITERATOR);
+                    if implements_trait(cx, cx.tables.expr_ty(iter_expr), iter_def_id, &[]);
```

Comment:
Just to make sure I understand correctly: this is checking that the scrutinee is not calling an additional `iter` method, right?

Reply:
It is checking, that a (method) call before the `next` call does not return a type, that implements iterator. Should I add a comment?

This does not cover this:

```rust
#[derive(Debug, Clone)]
struct S<'a> {
    chars: std::str::CharIndices<'a>,
}

while let Some(_) = s.clone().chars.next() {
    break;
}
```

But then again, if the struct is clonable, every field is clonable (in general), and this should be changed to `s.chars.clone().next()`, which prevents the lint from triggering.